controlplane/kubeadm/internal/machinefilters/machine_filters.go

```diff
@@ -197,3 +204,140 @@ func ControlPlaneSelectorForCluster(clusterName string) labels.Selector {
 		must(labels.NewRequirement(clusterv1.MachineControlPlaneLabelName, selection.Exists, []string{})),
 	)
 }
+
+// MatchesKCPConfiguration returns a filter to find all machines that matches with KCP config and do not require any rollout.
+// Kubernetes version, infrastructure template, and KubeadmConfig field need to be equivalent.
+func MatchesKCPConfiguration(ctx context.Context, c client.Client, kcp controlplanev1.KubeadmControlPlane, cluster clusterv1.Cluster) func(machine *clusterv1.Machine) bool {
+	return And(
+		MatchesKubernetesVersion(kcp.Spec.Version),
+		MatchesKubeadmBootstrapConfig(ctx, c, kcp, cluster),
+		MatchesTemplateClonedFrom(ctx, c, kcp),
+	)
+}
+
+// MatchesTemplateClonedFrom returns a filter to find all machines that match a given KCP infra template.
+func MatchesTemplateClonedFrom(ctx context.Context, c client.Client, kcp controlplanev1.KubeadmControlPlane) Func {
+	return func(machine *clusterv1.Machine) bool {
+		if machine == nil {
+			return false
+		}
+		infraObj, err := external.Get(ctx, c, &machine.Spec.InfrastructureRef, kcp.Namespace)
+		// Return true here because failing to get infrastructure machine should not be considered as unmatching.
+		if err != nil {
+			return true
```

Comment:
I'm trying to think about what types of scenarios we might hit with returning true here and what the end result may be. 

Is the worst case that we would fail to consider a Machine for upgrade for a particular reconciliation and expect that it would resolve itself on a following reconciliation?

Should we consider the case that the InfrastructureRef is missing (404/NotFound) rather than any error from the client and treat it separately? This might be something that falls under the banner of remediation rather than upgrade handling, though...

Reply:
+1 to be safe here, IIRC I added as similar comment somewhere else, when getting external references

Reply:
I didn't treat `InfrastructureRef is missing` case differently because as mentioned remediation may be handled somewhere else. For the rest of the errors, we do not want to rollout

Reply:
Yeah, makes sense. Remediation might handle this separately, if there are intermittent errors the safest thing to do is what's in the code today (ignore changes and continue); when reconciliation happens again it should reconcile and detect changes if any.

Reply:
@detiber thoughts?

Reply:
+1 agree, thanks for the clarification